xarray/core/common.py

```diff
@@ -591,6 +591,55 @@ def rolling(self, dim=None, min_periods=None, center=False, **dim_kwargs):
         return self._rolling_cls(self, dim, min_periods=min_periods,
                                  center=center)
 
+    def coarsen(self, dim=None, side='left', trim_excess=False,
+                coordinate_func=None, **dim_kwargs):
+        """
+        Coarsen object.
+
+        Parameters
+        ----------
+        dim: dict, optional
+            Mapping from the dimension name to the window size.
+        side : 'left' or 'right', or dict
+            If left, coarsening windows start from 0th index. The excessed
+            entries in the most right will be removed (if trim_excess is True).
+            If right, coarsen windows ends at the most right entry, while
+            excessed entries in the most left will be removed.
+        trim_excess : boolean, default False
```

Comment:
It would be nice to have an API that lets us express at least three options:
- trim excess entries
- pad with NaN
- raise an error if the shape does not divide exactly (this is probably the safest default behavior)

Maybe a string valued keyword argument would work better here, e.g., `boundary='trim'`, `boundary='pad'` and `boundary='exact'`?

I would also suggest putting this argument before `side`, since `side` is only used for a particular (non-default) value of this argument.